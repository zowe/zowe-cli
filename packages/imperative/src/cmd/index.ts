/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./doc/profiles/definition/ICommandProfileSchema";
export * from "./doc/profiles/definition/ICommandProfileProperty";
export * from "./doc/profiles/definition/ICommandProfileTypeConfiguration";
export * from "./doc/handler/ICommandHandler";
export * from "./doc/handler/ICommandHandlerResponseChecker";
export * from "./doc/handler/ICommandHandlerResponseValidator";
export * from "./doc/handler/IHandlerParameters";
export * from "./doc/handler/IChainedHandlerEntry";
export * from "./doc/handler/IChainedHandlerArgumentMapping";

export * from "./doc/option/ICommandOptionAllowableValues";
export * from "./doc/option/ICommandOptionDefinition";
export * from "./doc/option/ICommandOptionValueImplications";
export * from "./doc/option/ICommandPositionalDefinition";

export * from "./doc/response/response/ICommandResponse";
export * from "./doc/response/parms/ICommandResponseParms";
export * from "./doc/response/response/ICommandValidatorError";
export * from "./doc/response/response/ICommandValidatorResponse";

export * from "./doc/ICommandDefinition";
export * from "./doc/ICommandDefinitionPassOn";
export * from "./doc/ICommandExampleDefinition";
export * from "./doc/IPartialCommandDefinition";

export * from "./doc/args/ICommandArguments";

export * from "./handlers/FailedCommandHandler";

export * from "./help/abstract/AbstractHelpGenerator";
export * from "./help/abstract/AbstractHelpGeneratorFactory";

export * from "./help/doc/IHelpGeneratorFactory";
export * from "./help/doc/IHelpGeneratorFactoryParms";
export * from "./help/doc/IHelpGeneratorParms";
export * from "./help/DefaultHelpGenerator";
export * from "./help/HelpConstants";
export * from "./help/HelpGeneratorFactory";
export * from "./help/doc/IHelpGenerator";
export * from "./help/WebHelpGenerator";
export * from "./help/WebHelpManager";

export * from "./doc/profiles/definition/ICommandProfile";
export * from "./doc/profiles/definition/ICommandProfileTypeConfiguration";

export * from "./profiles/CliProfileManager";

export * from "./syntax/SyntaxValidator";

export * from "./utils/CommandUtils";
export * from "./utils/SharedOptions";

export * from "./yargs/doc/IYargsParms";
export * from "./yargs/doc/IYargsResponse";

export * from "./yargs/AbstractCommandYargs";
export * from "./yargs/CommandYargs";
export * from "./yargs/GroupCommandYargs";
export * from "./yargs/YargsConfigurer";
export * from "./yargs/YargsDefiner";

export * from "./CommandPreparer";
export * from "./CommandProcessor";
export * from "./response/CommandResponse";

export * from "./profiles/CommandProfiles";

export * from "./response/CommandResponse";
export * from "./response/HandlerResponse";

export * from "./doc/response/api/handler/IHandlerResponseApi";
export * from "./doc/response/api/handler/IHandlerProgressApi";
export * from "./doc/response/api/handler/IHandlerResponseConsoleApi";
export * from "./doc/response/api/handler/IHandlerResponseDataApi";
export * from "./doc/response/api/handler/IHandlerFormatOutputApi";

export * from "./doc/response/response/ICommandOutputFormat";
