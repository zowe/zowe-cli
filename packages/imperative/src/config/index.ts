/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./Config";
export * from "./ConfigAutoStore";
export * from "./ConfigConstants";
export * from "./ConfigSchema";
export * from "./ConfigBuilder";
export * as ConfigUtils from "./ConfigUtils";
export * from "./ProfileCredentials";
export * from "./ProfileInfo";
export * from "./ProfInfoErr";
export * from "./doc/IConfig";
export * from "./doc/IConfigAutoStoreOpts";
export * from "./doc/IConfigBuilderOpts";
export * from "./doc/IConfigConvertResult";
export * from "./doc/IConfigLayer";
export * from "./doc/IConfigOpts";
export * from "./doc/IConfigProfile";
export * from "./doc/IConfigSchema";
export * from "./doc/IConfigSecure";
export * from "./doc/IConfigVault";
export * from "./doc/IProfArgAttrs";
export * from "./doc/IProfAttrs";
export * from "./doc/IProfInfoUpdatePropOpts";
export * from "./doc/IProfInfoRemoveKnownPropOpts";
export * from "./doc/IProfInfoErrParms";
export * from "./doc/IProfLoc";
export * from "./doc/IProfMergeArgOpts";
export * from "./doc/IProfMergedArg";
export * from "./doc/IProfInfoProps";
export * from "./doc/IProfOpts";
