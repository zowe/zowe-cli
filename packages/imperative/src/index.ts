/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./censor";
export * from "./cmd";
export * from "./config";
export * from "./console";
export * from "./constants";
export * from "./events";
export * from "./error";
export * from "./expect";
export * from "./imperative";
// export * from "./interfaces";
export * from "./io";
export * from "./logger";
export * from "./messages";
export * from "./operations";
export * from "./profiles";
export * from "./rest";
export * from "./security";
export * from "./settings";
export * from "./utilities";
