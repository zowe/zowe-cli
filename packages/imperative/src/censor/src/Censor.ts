/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import * as lodash from "lodash";
import { Arguments } from "yargs";
import { ICommandProfileProperty } from "../../cmd/src/doc/profiles/definition/ICommandProfileProperty";
import { CliUtils } from "../../utilities/src/CliUtils";
import { ICensorOptions } from "./doc/ICensorOptions";
import { Config } from "../../config/src/Config";
import { ImperativeConfig } from "../../utilities/src/ImperativeConfig";
import { EnvironmentalVariableSettings } from "../../imperative/src/env/EnvironmentalVariableSettings";
import { ICommandProfileTypeConfiguration } from "../../cmd/src/doc/profiles/definition/ICommandProfileTypeConfiguration";
import { IProfileSchema} from "../../profiles/src/doc/definition/IProfileSchema";
import { IProfileTypeConfiguration } from "../../profiles/src/doc/config/IProfileTypeConfiguration";

export class Censor {

    /*********************************************************************
    * Basic censorship items - list definitions & initialiazations, etc. *
    **********************************************************************/

    /*
    * NOTE(Kelosky): Ideally we might have a consolidated list for secure fields, but for now we'll just
    * make sure they're collocated within the same class.
    *
    * NOTE(Harn): This list should be kept in sync with the base profile secure definitions and MUST be in camel case.
    */
    private static readonly MAIN_CENSORED_OPTIONS = ["auth", "authentication", "basicAuth", "certFilePassphrase", "credentials",
        "pw", "pass", "password", "passphrase", "tv", "tokenValue"];

    private static readonly MAIN_SECURE_PROMPT_OPTIONS = ["keyPassphrase", "password", "passphrase", "tokenValue", "user"];

    // The censor response.
    public static readonly CENSOR_RESPONSE = "****";

    // A set of default censored options.
    public static get DEFAULT_CENSORED_OPTIONS(): string[] {
        const censoredList = new Set<string>();
        for (const option of this.MAIN_CENSORED_OPTIONS) {
            censoredList.add(option);
            censoredList.add(CliUtils.getOptionFormat(option).kebabCase);
        }
        return Array.from(censoredList);
    }

    // Return a customized list of secure prompt options
    public static get SECURE_PROMPT_OPTIONS(): string[] {
        const censoredList = new Set<string>();
        for (const option of this.MAIN_SECURE_PROMPT_OPTIONS) {
            censoredList.add(option);
            censoredList.add(CliUtils.getOptionFormat(option).kebabCase);
        }
        return Array.from(censoredList);
    }

    // Set a censored options list that can be set and retrieved for each command.
    private static mCensoredOptions: Set<string> = new Set(this.DEFAULT_CENSORED_OPTIONS);

    // Return a customized list of censored options (or just the defaults if not set).
    public static get CENSORED_OPTIONS(): string[] {
        return Array.from(this.mCensoredOptions);
    }

    //Singleton caches of the Config, Command Definition and Command Arguments
    private static mConfig: Config = null;

    /**
     * Singleton implementation of an internal reference to the schema
     */
    private static mSchema: ICommandProfileTypeConfiguration[] = null;

    /**
     * Helper method to get an internal reference to the loaded profiles
     */
    public static get profileSchemas(): ICommandProfileTypeConfiguration[] {
        if (this.mSchema == null) this.mSchema = ImperativeConfig.instance.loadedConfig?.profiles ?? [];
        return this.mSchema;
    }

    /**
     * Helper method to set an internal reference to loaded profiles
     * @param _schemas - The schmas to pass in to set to the logger
     */
    public static setProfileSchemas(_schemas: IProfileTypeConfiguration[] | Map<string, IProfileSchema>) {
        if (this.mSchema == null) {
            this.mSchema = [];
        }
        if (_schemas instanceof Map) {
            _schemas.forEach((v: IProfileSchema) => {
                this.mSchema.push({ type: v.type, schema: v });
            });
        } else if (Array.isArray(_schemas)) {
            _schemas.forEach((v: IProfileTypeConfiguration) => {
                this.mSchema.push({ type: v.type, schema: v.schema });
            });
        }
    }

    /****************************************************
     * Helper functions for more advanced functionality *
     ****************************************************/

    /**
     * Helper function to handle profile schemas when setting the censored options
     * @param {IProfileTypeConfiguration | ICommandProfileTypeConfiguration} profileType - the profile type configuration to iterate over
     */
    private static handleSchema(profileType: IProfileTypeConfiguration | ICommandProfileTypeConfiguration): void {
        const secureOptions: Set<string> = new Set();

        /* eslint-disable-next-line no-unused-vars */
        for (const [key, cmdProp] of Object.entries(profileType.schema.properties)) {
            const prop = cmdProp as ICommandProfileProperty;
            // Add censored options from the schema if the option is secure
            if (prop.secure) {
                // Handle the case of a single option definition
                if (prop.optionDefinition) {
                    secureOptions.add(prop.optionDefinition.name);
                    for (const alias of prop.optionDefinition.aliases || []) {
                        // Remember to add the alias
                        secureOptions.add(alias);
                    }
                } else if (prop.optionDefinitions) {
                    // Handle the case of multiple option definitions
                    prop.optionDefinitions.forEach(opDef => {
                        secureOptions.add(opDef.name);
                        for (const alias of opDef.aliases || []) {
                            // Remember to add the alias
                            secureOptions.add(alias);
                        }
                    });
                } else {
                    secureOptions.add(key);
                }
            }
        }

        secureOptions.forEach(prop => this.addCensoredOption(prop));
    }

    /**
     * Add a censored option, including it's camelCase and kebabCase versions
     * @param {string} option - The option to censor
     */
    private static addCensoredOption(option: string) {
        // This option is required, but we do not want to ever allow null or undefined itself into the censored options
        if (option != null) {
            this.mCensoredOptions.add(option);
            this.mCensoredOptions.add(CliUtils.getOptionFormat(option).camelCase);
            this.mCensoredOptions.add(CliUtils.getOptionFormat(option).kebabCase);
        }
    }

    /**
     * Specifies whether a given property path (e.g. "profiles.lpar1.properties.host") is a special value or not.
     * Special value: Refers to any value defined as secure in the schema definition.
     *                These values should be already masked by the application (and/or plugin) developer.
     * @param prop Property path to determine if it is a special value
     * @returns True - if the given property is to be treated as a special value; False - otherwise
     */
    public static isSpecialValue(prop: string): boolean {
        let specialValues = this.SECURE_PROMPT_OPTIONS;
        const getPropertyNames = (prop: ICommandProfileProperty): string[] => {
            const ret: string[] = [];
            ret.push(prop.optionDefinition?.name);
            prop.optionDefinitions?.map(opDef => ret.push(opDef.name));
            return ret;
        };

        for (const profile of this.profileSchemas) {
            for (const prop of Object.values(profile.schema.properties)) {
                if (prop.secure) specialValues = lodash.union(specialValues, getPropertyNames(prop));
            }
        }

        return specialValues.some((v) => prop.endsWith(`.${v}`));
    }

    /**
     * Identifies if a property is a secure property
     * @param {string} prop - The property to check
     * @returns {boolean} - True if the property is secure; False otherwise
     */
    public static isSecureValue(prop: string) {
        return this.CENSORED_OPTIONS.includes(prop);
    }

    /****************************************************************************************
     * Bread and butter functions, setting up the class and performing censorship of values *
     ****************************************************************************************/

    /**
     * Generate and set the list of censored options.
     * Attempt to source the censored options from the schema, config, and/or command being executed.
     * @param {ICensorOptions} censorOpts - The objects to use to gather options that should be censored
     */
    public static setCensoredOptions(censorOpts?: ICensorOptions) {
        this.mCensoredOptions = new Set(this.DEFAULT_CENSORED_OPTIONS);

        if (censorOpts) {
            // Save off the config object
            this.mConfig = censorOpts.config;

            // If we have a ProfileTypeConfiguration (i.e. ImperativeConfig.instance.loadedConfig.profiles)
            if (censorOpts.profiles) { this.setProfileSchemas(censorOpts.profiles); }

            for (const profileType of this.profileSchemas ?? []) {
                // If we know the command we are running, and we know the profile types that the command uses
                // we should only use those profiles to determine what should be censored. If we do not, we should
                // add everything
                if (censorOpts.commandDefinition == null ||
                    censorOpts.commandDefinition.profile?.optional?.includes(profileType.type) ||
                    censorOpts.commandDefinition.profile?.required?.includes(profileType.type)) {
                    this.handleSchema(profileType);
                }
            }

            // Include any secure options from the config
            if (censorOpts.config) {
                const secureOptions: Set<string> = new Set();

                // Try to use the command and inputs to find the profiles being loaded
                if (censorOpts.commandDefinition && censorOpts.commandArguments) {
                    const profiles = [];
                    for (const prof of censorOpts.commandDefinition.profile?.required || []) {
                        profiles.push(prof);
                    }
                    for (const prof of censorOpts.commandDefinition.profile?.optional || []) {
                        profiles.push(prof);
                    }

                    for (const prof of profiles) {
                        // If the profile exists, append all of the secure props to the censored list
                        let profName = censorOpts.commandArguments?.[`${prof}-profile`];
                        if (!profName) { profName = this.mConfig.mProperties.defaults[`${prof}`]; }
                        if (profName && censorOpts.config.api.profiles.get(profName)) {
                            censorOpts.config.api.secure.securePropsForProfile(profName).forEach(prop => secureOptions.add(prop));
                        }
                    }
                } else {
                    // We only have a configuration file, assume every property that is secured should be censored
                    censorOpts.config.api.secure.findSecure(censorOpts.config.mProperties.profiles, "profiles").forEach(
                        prop => secureOptions.add(prop.split(".").pop())
                    );
                }
                secureOptions.forEach(prop => this.addCensoredOption(prop));
            }
        } else if (this.profileSchemas) {
            for (const profileType of this.profileSchemas) {
                this.handleSchema(profileType);
            }
        }
    }

    /**
     * Copy and censor any sensitive CLI arguments before logging/printing
     * @param {string[]} args - The args list to censor
     * @returns {string[]}
     */
    public static censorCLIArgs(args: string[]): string[] {
        const newArgs: string[] = JSON.parse(JSON.stringify(args));
        const censoredValues = this.CENSORED_OPTIONS.map(CliUtils.getDashFormOfOption);
        for (const value of censoredValues) {
            if (args.indexOf(value) >= 0) {
                const valueIndex = args.indexOf(value);
                if (valueIndex < args.length - 1) {
                    newArgs[valueIndex + 1] = this.CENSOR_RESPONSE; // censor the argument after the option name
                }
            }
        }
        return newArgs;
    }

    /**
     * Copy and censor any sensitive CLI arguments before logging/printing
     * @param {string} data - the data to censor
     * @returns {string} - the censored data
     */
    public static censorRawData(data: string, category: string = ""): string {
        const config = this.mConfig ?? ImperativeConfig.instance?.config;

        // Return the data if not using config
        if (!config?.exists) return data;

        // Return the data if we are printing to the console and masking is disabled
        if (ImperativeConfig.instance?.envVariablePrefix) {
            const envMaskOutput = EnvironmentalVariableSettings.read(ImperativeConfig.instance.envVariablePrefix).maskOutput.value;
            const envShowSecureArgs = EnvironmentalVariableSettings.read(ImperativeConfig.instance.envVariablePrefix).showSecureArgs.value;
            // Hardcoding "console" instead of using Logger.DEFAULT_CONSOLE_NAME because of circular dependencies
            if ((category === "console" || category === "json") &&
                (envMaskOutput.toUpperCase() === "FALSE" || envShowSecureArgs.toUpperCase() === "TRUE")) {
                return data;
            }
        }

        let newData = data;

        const secureFields = config.api.secure.findSecure(config.mProperties.profiles, "profiles");
        for (const prop of secureFields) {
            const sec = lodash.get(config.mProperties, prop);
            if (sec && typeof sec !== "object" && !this.isSpecialValue(prop) && this.isSecureValue(prop.split(".").pop())) {
                newData = newData.replace(new RegExp(sec, "gi"), this.CENSOR_RESPONSE);
            }
        }
        return newData;
    }

    /**
     * Copy and censor a yargs argument object before logging
     * @param {yargs.Arguments} args - the args to censor
     * @returns {yargs.Arguments} - a censored copy of the arguments
     */
    public static censorYargsArguments(args: Arguments): Arguments {
        const newArgs: Arguments = JSON.parse(JSON.stringify(args));

        for (const optionName of Object.keys(newArgs)) {
            if (this.CENSORED_OPTIONS.indexOf(optionName) >= 0) {
                const valueToCensor = newArgs[optionName];
                newArgs[optionName] = this.CENSOR_RESPONSE;
                for (const checkAliasKey of Object.keys(newArgs)) {
                    if (newArgs[checkAliasKey] === valueToCensor) {
                        newArgs[checkAliasKey] = this.CENSOR_RESPONSE;
                    }
                }
            }
        }
        return newArgs;
    }
}