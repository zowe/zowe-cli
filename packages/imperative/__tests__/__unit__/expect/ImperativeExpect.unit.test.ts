/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ImperativeExpect } from "../../../src/expect";
import { ImperativeError } from "../../../src/error";
import { TestLogger } from "../../../__tests__/src/TestLogger";
// UnitTestUtils.replaceIt();

const obj = {
    data: "something",
    isReal: "yesIAm",
    blank: "",
};

const nestedObj = {
    data: "something",
    isReal: "yesIAm",
    blank: "",
    nest: {
        prop1: "hello",
        prop2: false,
        theZeroLengthArray: [] as any,
        containsArray: ["hello"]
    }
};

describe("ImperativeExpect tests", () => {
    describe("toMatchRegExp", () => {
        it("should not throw an error if the value matches the provided regular expression", () => {
            let error: ImperativeError = {} as any;
            try {
                ImperativeExpect.toMatchRegExp("token", "^token$");
            } catch(thrownError) {
                error = thrownError;
            }
            expect(error).toEqual({});
        });
        it("should throw an error if the value does not match the provided regular expression with custom message", () => {
            let error: ImperativeError = {} as any;
            try {
                ImperativeExpect.toMatchRegExp("token", "^token1", "test");
            } catch(thrownError) {
                error = thrownError;
            }
            expect(error.message).toContain("test");
        });
    });

    it("Should throw an error for an undefined key when we expect it to be defined", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeDefined(obj, ["noExist"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("Should not throw an error for a list of defined keys", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeDefined(obj, ["data", "isReal", "blank"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("Should throw an error for a defined key when the last one is undefined", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeDefined(obj, ["data", "isReal", "notReal"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("Should not throw an error for a non-blank field", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.toBeDefinedAndNonBlank("hey", "not_blank");
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("Should throw an error for a blank field", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.toBeDefinedAndNonBlank(" ", "blank");
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("Should throw an error for a defined key that is blank", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeDefinedAndNonBlank(obj, ["blank"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("Should not throw an error for a defined key that is blank", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeDefined(obj, ["blank"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("Should not throw an error if the item is found in a list", () => {
        const exists = "exists";
        expect(ImperativeExpect.toBeOneOf(exists, [exists, "one", "two"])).toBe(exists);
        expect(ImperativeExpect.toBeOneOf(exists, ["one", "two", exists])).toBe(exists);
        expect(ImperativeExpect.toBeOneOf(exists, ["one", exists, "two"])).toBe(exists);
    });

    it("Should throw an error if an item is not found in a list", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeUndefined(obj, ["noExist"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("Should not throw an error for an undefined key when we expect undefined", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeUndefined(obj, ["noExist"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("Should throw an error for defined key that we want to be undefined", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeUndefined(obj, ["data"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("Should throw an error for defined key that is last in a list of undefined keys", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeUndefined(obj, ["fake1", "fake2", "data"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should not throw error if the type is correct", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeOfType(obj, "string", ["data"]);
        } catch (thrownError) {
            error = thrownError;
            expect(0).toBe(1);
        }
        expect(error).toBeUndefined();
    });

    it("should throw an error if the type is incorrect", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeOfType(obj, "boolean", ["data"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should not throw error if the type is correct of a nested property", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeOfType(nestedObj, "boolean", ["nest.prop2"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("should throw error if the type is not correct of a nested property", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeOfType(nestedObj, "string", ["nest.prop2"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should throw error if the input is not an array and an array is expected", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeAnArray(nestedObj, false, ["nest.prop2"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should throw error if the input array is of length 0", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeAnArray(nestedObj, true, ["nest.theZeroLengthArray"]);
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should not throw error if the input array is of length 0", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.keysToBeAnArray(nestedObj, false, ["nest.theZeroLengthArray"]);
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("should not throw error if the entry is found", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.arrayToContain(nestedObj.nest.containsArray, (entry) => {
                return entry === "hello";
            });
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("should throw an error if the entry is NOT found", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.arrayToContain(nestedObj.nest.containsArray, (entry) => {
                return entry === "test";
            });
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should allow an expect failure to be transformed", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.expectAndTransform(() => {
                ImperativeExpect.toBeEqual(0, 1);
            }, (impErr) => {
                return new ImperativeError({ msg: "The values weren't equal!!!" });
            });
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should allow a custom error message", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.toBeEqual(0, 1, "This is custom");
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error.message).toMatchSnapshot();
    });

    it("should not throw error if the values are not equal", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.toNotBeEqual("this", "that");
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("should throw an error if the values are equal and we're expected them to not be equal", () => {
        let error: ImperativeError;
        try {
            ImperativeExpect.toNotBeEqual("this", "this");
        } catch (thrownError) {
            error = thrownError;
        }
        TestLogger.info(error.message);
        expect(error).toMatchSnapshot();
    });
});
