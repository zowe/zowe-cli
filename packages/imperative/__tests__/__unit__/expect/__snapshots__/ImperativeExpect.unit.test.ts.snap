// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImperativeExpect tests Should throw an error for a blank field 1`] = `"Expect Error: Required parameter 'blank' must not be blank"`;

exports[`ImperativeExpect tests Should throw an error for a defined key that is blank 1`] = `"Expect Error: Required parameter 'blank' must not be blank"`;

exports[`ImperativeExpect tests Should throw an error for a defined key when the last one is undefined 1`] = `"Expect Error: Required parameter 'notReal' must be defined"`;

exports[`ImperativeExpect tests Should throw an error for an undefined key when we expect it to be defined 1`] = `"Expect Error: Required parameter 'noExist' must be defined"`;

exports[`ImperativeExpect tests Should throw an error for defined key that is last in a list of undefined keys 1`] = `"Expect Error: Required parameter 'data' must be undefined"`;

exports[`ImperativeExpect tests Should throw an error for defined key that we want to be undefined 1`] = `"Expect Error: Required parameter 'data' must be undefined"`;

exports[`ImperativeExpect tests should allow a custom error message 1`] = `"Expect Error: This is custom"`;

exports[`ImperativeExpect tests should allow an expect failure to be transformed 1`] = `"The values weren't equal!!!"`;

exports[`ImperativeExpect tests should throw an error if the entry is NOT found 1`] = `"Expect Error: The required entry was NOT found within the input array: 'hello'"`;

exports[`ImperativeExpect tests should throw an error if the type is incorrect 1`] = `"Expect Error: Object key 'data' must be of type 'boolean'"`;

exports[`ImperativeExpect tests should throw an error if the values are equal and we're expected them to not be equal 1`] = `[Error: Expect Error: Values specified are equal ("this")]`;

exports[`ImperativeExpect tests should throw error if the input array is of length 0 1`] = `"Expect Error: Required parameter 'nest.theZeroLengthArray' must be an array with at least 1 entry."`;

exports[`ImperativeExpect tests should throw error if the input is not an array and an array is expected 1`] = `"Expect Error: Required parameter 'false' must be an array."`;

exports[`ImperativeExpect tests should throw error if the type is not correct of a nested property 1`] = `"Expect Error: Object key 'nest.prop2' must be of type 'string'"`;
